include/deal.II/base/array_view.h

```diff
@@ -252,35 +259,97 @@ class ArrayView
    */
   const std::size_t n_elements;
 
-  friend class ArrayView<const ElementType>;
+  friend class ArrayView<const ElementType, MemorySpaceType>;
 };
 
 
 
 //---------------------------------------------------------------------------
 
 
+namespace internal
+{
+  namespace ArrayViewHelper
+  {
+    template <typename MemorySpaceType>
+    inline bool
+    correct_memory_space(const void *const)
+    {
+      return std::is_same<MemorySpaceType, MemorySpace::Host>::value;
+    }
+
+#ifdef DEAL_II_COMPILER_CUDA_AWARE
+    template <>
+    inline bool
+    correct_memory_space<MemorySpace::Host>(const void *ptr)
+    {
+      cudaPointerAttributes attributes;
+      const cudaError_t cuda_error = cudaPointerGetAttributes(&attributes, ptr);
+      if (cuda_error != cudaErrorInvalidValue)
+        {
+          AssertCuda(cuda_error);
+          return attributes.memoryType == cudaMemoryTypeHost;
+        }
+      else
+        {
+          // ignore and reset the error since host pointers produce an error
+          cudaGetLastError();
+          return true;
+        }
+    }
 
-template <typename ElementType>
-inline ArrayView<ElementType>::ArrayView(value_type *      starting_element,
-                                         const std::size_t n_elements)
+    template <>
+    inline bool
+    correct_memory_space<MemorySpace::CUDA>(const void *ptr)
```

Comment:
it looks like you could relatively easily combine this and the previous function by using `std::is_same_type` in one place. I think that would make things a bit easier to read.

Reply:
I am always a bit afraid to have `DEAL_II_COMPILER_CUDA_AWARE` inside a function that is potentially instantiated by the host compiler. Guarding with `static_assert` should be sufficient though.